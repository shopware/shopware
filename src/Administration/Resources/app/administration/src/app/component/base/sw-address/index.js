import template from './sw-address.html.twig';
import './sw-address.scss';

const { Component } = Shopware;

/**
 * @public
 * @description Component to render a postal address
 * @status ready
 * @example-type static
 * @component-example
 * <sw-address headline="Billing address" :address="{
 *     salutation: 'Mister',
 *     title: 'Doctor',
 *     firstName: 'John',
 *     lastName: 'Doe',
 *     street: 'Main St 123',
 *     zipcode: '12456',
 *     city: 'Anytown',
 *     country: { name: 'Germany' }
 * }"></sw-address>
 */
Component.register('sw-address', {
    template,

    props: {
        address: {
            type: Object,
            default() {
                return {
                    salutation: '',
                    title: '',
                    firstName: '',
                    lastName: '',
                    street: '',
                    zipcode: '',
                    city: '',
                    country: {
                        name: '',
                    },
                    countryState: {
                        name: '',
                    },
                };
            },
        },

        headline: {
            type: String,
            required: false,
            default: '',
        },

        showEditButton: {
            type: Boolean,
            required: false,
            default: false,
        },

        editLink: {
            type: Object,
            required: false,
            default: null,
        },
    },

    computed: {
        addressClasses() {
            return {
                'sw-address--headline': this.headline,
            };
        },
    },
});
