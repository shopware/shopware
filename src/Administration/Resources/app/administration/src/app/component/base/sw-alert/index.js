import template from './sw-alert.html.twig';
import './sw-alert.scss';

const { Component } = Shopware;

/**
 * @description
 * The <u>sw-alert</u> component is used to convey important information to the user. It comes in 4 variations,
 * <strong>success</strong>, <strong>info</strong>, <strong>warning</strong> and <strong>error</strong>. These have
 * default icons assigned which can be changed and represent different actions
 * @status ready
 * @example-type dynamic
 * @component-example
 * <sw-alert variant="info" title="Example title" :closable="true">
 *    Sample text
 * </sw-alert>
 */
Component.register('sw-alert', {
    template,

    props: {
        variant: {
            type: String,
            default: 'info',
            validValues: ['info', 'warning', 'error', 'success'],
            validator(value) {
                return ['info', 'warning', 'error', 'success'].includes(value);
            }
        },
        appearance: {
            type: String,
            default: 'default',
            validValues: ['default', 'notification', 'system'],
            validator(value) {
                return ['default', 'notification', 'system'].includes(value);
            }
        },
        title: {
            type: String,
            required: false,
            default: ''
        },
        showIcon: {
            type: Boolean,
            required: false,
            default: true
        },
        closable: {
            type: Boolean,
            required: false,
            default: false
        },
        notificationIndex: {
            type: String,
            required: false
        }
    },
    computed: {
        alertClasses() {
            return [
                `sw-alert--${this.variant}`,
                `sw-alert--${this.appearance}`,
                {
                    'sw-alert--no-icon': !this.showIcon,
                    'sw-alert--closable': this.closable,
                    'sw-alert--actions': this.hasActionSlot
                }
            ];
        },

        alertIcon() {
            const iconConfig = {
                info: 'default-badge-info',
                warning: 'default-badge-warning',
                error: 'default-badge-error',
                success: 'default-basic-checkmark-circle'
            };

            return iconConfig[this.variant] || 'default-bell-bell';
        },

        hasActionSlot() {
            return !!this.$slots.actions;
        }
    }
});
