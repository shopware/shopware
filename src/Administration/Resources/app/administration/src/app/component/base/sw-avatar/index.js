import template from './sw-avatar.html.twig';
import './sw-avatar.scss';

const { Component } = Shopware;

const colors = [
    '#FFD700',
    '#FFC700',
    '#F88962',
    '#F56C46',
    '#FF85C2',
    '#FF68AC',
    '#6AD6F0',
    '#4DC6E9',
    '#A092F0',
    '#8475E9',
    '#57D9A3',
    '#3CCA88'
];

/**
 * @description The component helps adding a custom user image or initials to the administration.
 * @status ready
 * @example-type static
 * @component-example
 * <div style="display: flex; align-items: center;">
 * <sw-avatar color="#dd4800"
 *            size="48px"
 *            firstName="John"
 *            style="margin: 0 10px;"
 *            lastName="Doe"></sw-avatar>
 *
 * <sw-avatar size="48px"
 *            imageUrl="https://randomuser.me/api/portraits/women/68.jpg"></sw-avatar>
 * </div>
 */
Component.register('sw-avatar', {
    template,

    props: {
        color: {
            type: String,
            required: false,
            default: ''
        },
        size: {
            type: String,
            required: false
        },
        firstName: {
            type: String,
            required: false,
            default: ''
        },
        lastName: {
            type: String,
            required: false,
            default: ''
        },
        imageUrl: {
            type: String,
            required: false
        },
        placeholder: {
            type: Boolean,
            required: false,
            default: false
        }
    },

    data() {
        return {
            fontSize: 16,
            lineHeight: 16
        };
    },

    watch: {
        size() {
            this.$nextTick(() => {
                this.generateAvatarInitialsSize();
            });
        }
    },

    computed: {
        avatarSize() {
            const size = this.size;

            return {
                width: size,
                height: size
            };
        },

        avatarInitials() {
            const firstNameLetter = this.firstName ? this.firstName[0] : '';
            const lastNameLetter = this.lastName ? this.lastName[0] : '';

            return firstNameLetter + lastNameLetter;
        },

        avatarInitialsSize() {
            return {
                'font-size': `${this.fontSize}px`,
                'line-height': `${this.lineHeight}px`
            };
        },

        avatarImage() {
            if (!this.imageUrl) {
                return '';
            }

            return {
                'background-image': `url('${this.imageUrl}')`
            };
        },

        avatarColor() {
            if (this.color.length) {
                return {
                    'background-color': this.color
                };
            }

            const firstNameLength = this.firstName ? this.firstName.length : 0;
            const lastNameLength = this.lastName ? this.lastName.length : 0;

            const nameLength = firstNameLength + lastNameLength;
            const color = colors[nameLength % colors.length];

            return {
                'background-color': color
            };
        },

        showPlaceholder() {
            return this.placeholder && (!this.imageUrl || !this.imageUrl.length);
        },

        showInitials() {
            return !this.placeholder && (!this.imageUrl || !this.imageUrl.length);
        }
    },

    mounted() {
        this.mountedComponent();
    },

    methods: {
        mountedComponent() {
            this.generateAvatarInitialsSize();
        },

        generateAvatarInitialsSize() {
            const avatarSize = this.$refs.swAvatar.offsetHeight;

            this.fontSize = Math.round(avatarSize * 0.4);
            this.lineHeight = Math.round(avatarSize * 0.98);
        }
    }
});
