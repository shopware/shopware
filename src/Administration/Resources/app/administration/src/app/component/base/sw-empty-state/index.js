import template from './sw-empty-state.html.twig';
import './sw-empty-state.scss';

const { Component } = Shopware;

/**
 * @private
 */
Component.register('sw-empty-state', {
    template,

    props: {
        title: {
            type: String,
            default: '',
            required: true
        },
        subline: {
            type: String,
            default: '',
            required: false
        },
        color: {
            type: String,
            default: '',
            required: false
        },
        icon: {
            type: String,
            default: '',
            required: false
        }
    },

    computed: {
        moduleColor() {
            return this.color || this.$route.meta.$module.color;
        },

        moduleDescription() {
            return this.subline || this.$tc(this.$route.meta.$module.description);
        },

        moduleIcon() {
            return this.icon || this.$route.meta.$module.icon;
        },

        hasActionSlot() {
            return !!this.$slots.actions;
        }
    }
});
