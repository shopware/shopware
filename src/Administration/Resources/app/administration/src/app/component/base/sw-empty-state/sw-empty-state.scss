@import "~scss/variables";

.sw-empty-state {
    position: absolute;
    display: flex;
    width: 100%;
    height: 100%;
    overflow-x: hidden;
    overflow-y: auto;

    .sw-empty-state__element {
        max-width: 600px;
        padding: 20px;
        margin: 0 auto;
        align-self: center;
        text-align: center;
    }

    .sw-empty-state__title {
        font-size: 25px;
        margin-bottom: 15px;
        color: $color-darkgray-200;
    }

    .sw-empty-state__description {
        color: $color-darkgray-200;
        margin-bottom: 15px;
        font-size: 14px;
    }

    .sw-empty-state__icon {
        margin-bottom: 20px;
    }
}
