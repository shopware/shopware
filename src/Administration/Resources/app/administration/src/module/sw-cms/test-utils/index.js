import { mount } from '@vue/test-utils';

function runGenericCmsTest(component) {
    beforeAll(async () => {
        await setupCmsEnvironment();
    });

    it('should be a Vue.js component', async () => {
        const wrapper = mount(component);
        expect(wrapper.vm).toBeTruthy();
    });
}

function runCmsBlockRegistryTest(config) {
    beforeAll(async () => {
        await setupCmsEnvironment();
        await import(config.import);
    });

    it('registers the component in the CMS block registry', async () => {
        const componentRegistry = Shopware.Component.getComponentRegistry();
        const cmsService = Shopware.Service('cmsService');

        expect(componentRegistry.has(config.component)).toBeTruthy();
        expect(componentRegistry.has(config.preview)).toBeTruthy();

        const blockConfig = cmsService.getCmsBlockConfigByName(config.name);

        expect(blockConfig.component).toBe(config.component);
        expect(blockConfig.previewComponent).toBe(config.preview);
    });
}

function runCmsElementRegistryTest(config) {
    beforeAll(async () => {
        await setupCmsEnvironment();
        await import(config.import);
    });

    it('registers the component in the CMS block registry', async () => {
        const componentRegistry = Shopware.Component.getComponentRegistry();
        const cmsService = Shopware.Service('cmsService');

        expect(componentRegistry.has(config.component)).toBeTruthy();

        if (config.preview) {
            expect(componentRegistry.has(config.preview)).toBeTruthy();
        }

        const elementConfig = cmsService.getCmsElementConfigByName(config.name);

        expect(elementConfig.component).toBe(config.component);
        expect(elementConfig.configComponent).toBe(config.config);

        if (config.preview) {
            expect(elementConfig.previewComponent).toBe(config.preview);
        }
    });
}

async function setupCmsEnvironment() {
    await import('src/module/sw-cms/store/cms-page.store');
    await import('src/module/sw-cms/service/cms.service');
    await import('src/module/sw-cms/service/cms-element-favorites.service');
    await import('src/module/sw-cms/mixin/sw-cms-element.mixin');
    await import('src/module/sw-cms/mixin/sw-cms-state.mixin');

    Shopware.State.get('session').currentUser = {
        id: 'admin',
    };
}

/**
 * @private
 * @package buyers-experience
 */
export { runGenericCmsTest, runCmsBlockRegistryTest, runCmsElementRegistryTest, setupCmsEnvironment };
