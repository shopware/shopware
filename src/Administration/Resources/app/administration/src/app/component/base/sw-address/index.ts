import type { PropType } from 'vue';
import type { Route } from 'vue-router';
import type { Address } from 'src/core/service/api/custom-snippet.api.service';
import template from './sw-address.html.twig';
import './sw-address.scss';

const { Component } = Shopware;

/**
 * @package admin
 *
 * @deprecated tag:v6.6.0 - Will be private
 * @public
 * @description Component to render a postal address
 * @status ready
 * @example-type static
 * @component-example
 * <sw-address headline="Billing address" :address="{
 *     salutation: 'Mister',
 *     title: 'Doctor',
 *     firstName: 'John',
 *     lastName: 'Doe',
 *     street: 'Main St 123',
 *     zipcode: '12456',
 *     city: 'Anytown',
 *     country: { name: 'Germany' }
 * }" :formattingAddress="First Name Last Name\nGermany"></sw-address>
 */
Component.register('sw-address', {
    template,

    props: {
        address: {
            type: Object as PropType<Address>,
            default(): Address {
                return {
                    salutation: {
                        displayName: '',
                        translated: {
                            displayName: '',
                        },
                    },
                    title: '',
                    firstName: '',
                    lastName: '',
                    street: '',
                    zipcode: '',
                    city: '',
                    country: {
                        name: '',
                        translated: {
                            name: '',
                        },
                    },
                    countryState: {
                        name: '',
                        translated: {
                            name: '',
                        },
                    },
                };
            },
        },

        headline: {
            type: String,
            required: false,
            default: '',
        },

        formattingAddress: {
            type: String,
            required: false,
            default: null,
        },

        showEditButton: {
            type: Boolean,
            required: false,
            default: false,
        },

        editLink: {
            type: Object as PropType<Route | null>,
            required: false,
            default: null,
        },
    },

    computed: {
        addressClasses(): Record<string, boolean | string> {
            return {
                'sw-address--headline': this.headline,
            };
        },

        displayFormattingAddress(): string {
            return this.formattingAddress;
        },
    },
});
