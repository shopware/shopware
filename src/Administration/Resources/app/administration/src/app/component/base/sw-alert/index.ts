import type { NotificationType } from 'src/app/mixin/notification.mixin';
import type { PropType } from 'vue';
import template from './sw-alert.html.twig';
import './sw-alert.scss';

const { Component } = Shopware;
type AppearanceType = 'default' | 'notification' | 'system';
type CssClassesObject = { [key: string]: boolean };
type CssClasses = Array<string | CssClassesObject> | CssClassesObject;

/**
 * @package admin
 *
 * @deprecated tag:v6.6.0 - Will be private
 * @description
 * The <u>sw-alert</u> component is used to convey important information to the user. It comes in 4 variations,
 * <strong>success</strong>, <strong>info</strong>, <strong>warning</strong> and <strong>error</strong>. These have
 * default icons assigned which can be changed and represent different actions
 * @status ready
 * @example-type dynamic
 * @component-example
 * <sw-alert variant="info" title="Example title" :closable="true">
 *    Sample text
 * </sw-alert>
 */
Component.register('sw-alert', {
    template,

    props: {
        variant: {
            type: String as PropType<NotificationType>,
            required: false,
            default: 'info',
            validValues: ['info', 'warning', 'error', 'success', 'neutral'],
            validator(value: string): boolean {
                return ['info', 'warning', 'error', 'success', 'neutral'].includes(value);
            },
        },
        appearance: {
            type: String as PropType<AppearanceType>,
            required: false,
            default: 'default',
            validValues: ['default', 'notification', 'system'],
            validator(value: string) {
                return ['default', 'notification', 'system'].includes(value);
            },
        },
        title: {
            type: String,
            required: false,
            default: '',
        },
        showIcon: {
            type: Boolean,
            required: false,
            // TODO: Boolean props should only be opt in and therefore default to false
            // eslint-disable-next-line vue/no-boolean-default
            default: true,
        },
        closable: {
            type: Boolean,
            required: false,
            default: false,
        },
        notificationIndex: {
            type: String,
            required: false,
            default: null,
        },
        icon: {
            type: String,
            required: false,
            default: null,
        },
    },
    computed: {
        alertIcon(): string {
            if (this.icon) {
                return this.icon;
            }

            const iconConfig: { [type: string]: string } = {
                info: 'regular-info-circle',
                warning: 'regular-exclamation-triangle',
                error: 'regular-exclamation-circle',
                success: 'regular-check-circle',
                neutral: 'regular-info-circle',
            };

            return iconConfig[this.variant] || 'regular-bell';
        },

        hasActionSlot(): boolean {
            return !!this.$slots.actions;
        },

        alertClasses(): CssClasses {
            const variantClass = `sw-alert--${this.variant}`;

            return [
                `sw-alert--${this.appearance}`,
                {
                    [variantClass]: this.appearance !== 'system',
                    'sw-alert--icon': this.showIcon,
                    'sw-alert--no-icon': !this.showIcon,
                    'sw-alert--closable': this.closable,
                    'sw-alert--actions': this.hasActionSlot,
                },
            ];
        },

        alertBodyClasses(): CssClasses {
            return {
                'sw-alert__body--icon': this.showIcon,
                'sw-alert__body--closable': this.closable,
            };
        },
    },
});
