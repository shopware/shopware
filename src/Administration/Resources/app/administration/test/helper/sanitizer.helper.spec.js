import { shallowMount, createLocalVue } from '@vue/test-utils';

import Sanitizer from 'src/core/helper/sanitizer.helper';
import 'src/app/component/base/sw-empty-state';
import SanitizePlugin from 'src/app/plugin/sanitize.plugin';

// Disable developer hints in jest output
jest.spyOn(global.console, 'warn').mockImplementation(() => jest.fn());

describe('core/helper/sanitizer.helper.js', () => {
    // See for payload list: https://github.com/s0md3v/AwesomeXSS
    it('should sanitize the html', () => {
        expect(Sanitizer.sanitize('<A/hREf="j%0aavas%09cript%0a:%09con%0afirm%0d``">z'))
            .toBe('<a href="j%0aavas%09cript%0a:%09con%0afirm%0d``">z</a>');

        expect(Sanitizer.sanitize('<d3"<"/onclick="1>[confirm``]"<">z'))
            .toBe('z');

        expect(Sanitizer.sanitize('<d3/onmouseenter=[2].find(confirm)>z'))
            .toBe('z');

        expect(Sanitizer.sanitize('<d3/onmouseenter=[2].find(confirm)>z'))
            .toBe('z');

        expect(Sanitizer.sanitize('<details open ontoggle=confirm()>'))
            .toBe('<details open=""></details>');

        expect(Sanitizer.sanitize(`<script y="><">/*<script* */prompt()</script`)) // eslint-disable-line
            .toBe('');

        expect(Sanitizer.sanitize('<w="/x="y>"/ondblclick=`<`[confir\u006d``]>z'))
            .toBe('z');

        expect(Sanitizer.sanitize('<a href=javas&#99;ript:alert(1)>click'))
            .toBe('<a>click</a>');

        expect(Sanitizer.sanitize('<script/"<a"/src=data:=".<a,[8].some(confirm)>'))
            .toBe('');

        expect(Sanitizer.sanitize('<svg/x=">"/onload=confirm()//'))
            .toBe('');

        expect(Sanitizer.sanitize('<--`<img/src=` onerror=confirm``> --!>'))
            .toBe('&lt;--`<img src="`"> --!&gt;');

        expect(Sanitizer.sanitize('<svg%0Aonload=%09((pro\u006dpt))()//'))
            .toBe('');

        expect(Sanitizer.sanitize('<sCript x>(((confirm)))``</scRipt x>'))
            .toBe('');

        expect(Sanitizer.sanitize('<svg </onload ="1> (_=prompt,_(1)) "">'))
            .toBe('<svg></svg>');

        expect(Sanitizer.sanitize('<sCript x>(((confirm)))``</scRipt x>'))
            .toBe('');

        expect(Sanitizer.sanitize('<!--><script src=//14.rs>'))
            .toBe('');

        expect(Sanitizer.sanitize('<!--><script src=//14.rs>'))
            .toBe('');

        expect(Sanitizer.sanitize('<embed src=//14.rs>'))
            .toBe('');

        expect(Sanitizer.sanitize('<script x=">" src=//15.rs></script>'))
            .toBe('');

        expect(Sanitizer.sanitize('<!\'/*"/*/\'/*/"/*--></Script><Image SrcSet=K */; OnError=confirm`1` //>'))
            .toBe('<img srcset="K">');

        expect(Sanitizer.sanitize('<x oncut=alert()>x'))
            .toBe('x');

        expect(Sanitizer.sanitize('<svg onload=write()>'))
            .toBe('<svg></svg>');
    });

    it('should ensure a persistent configuration can be set and cleared', () => {
        const dirtyContent = '<my-component>abc</my-component>';

        expect(Sanitizer.sanitize(dirtyContent)).toBe('abc');
        Sanitizer.setConfig({
            ADD_TAGS: ['my-component']
        });
        expect(Sanitizer.sanitize(dirtyContent)).toBe('<my-component>abc</my-component>');
        Sanitizer.clearConfig();

        expect(Sanitizer.sanitize(dirtyContent)).toBe('abc');
    });

    it('should be able to modify the output using a middleware', () => {
        Sanitizer.addMiddleware('afterSanitizeElements', (node) => {
            if (node.nodeType && node.nodeType === document.TEXT_NODE) {
                node.textContent = 'foo';
            }
            return node;
        });
        const content = '<div><p>Beautiful City</p><p>Beautiful country</p></div>';
        const expected = '<div><p>foo</p><p>foo</p></div>';

        expect(Sanitizer.sanitize(content)).toBe(expected);
    });

    it('should register a middleware with a valid name only', () => {
        expect(Sanitizer.addMiddleware('foo', () => {})).toBe(false);
        expect(Sanitizer.addMiddleware('afterSanitizeElements', () => {})).toBe(true);

        expect(Sanitizer.removeMiddleware('afterSanitizeElements')).toBe(true);
    });

    it('should remove a middleware with a valid name only', () => {
        expect(Sanitizer.removeMiddleware('foo')).toBe(false);
        expect(Sanitizer.removeMiddleware('afterSanitizeElements')).toBe(true);
    });

    it('should sanitize untrusted HTML in a component', () => {
        const localVue = createLocalVue();
        localVue.use(SanitizePlugin);

        const $route = {
            meta: { $module: { icon: null } }
        };

        const $tc = (value) => {
            return value;
        };

        const wrapper = shallowMount(Shopware.Component.build('sw-empty-state'), {
            localVue,
            stubs: ['sw-icon'],
            mocks: {
                $route,
                $tc
            },
            props: {
                title: 'Foo bar',
                subline: '<x oncut=alert()>x'
            }
        });

        expect(wrapper.element).toMatchSnapshot();
    });
});
