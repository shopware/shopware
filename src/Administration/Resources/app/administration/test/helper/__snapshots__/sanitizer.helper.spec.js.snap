// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/helper/sanitizer.helper.js should sanitize untrusted HTML in a component 1`] = `
<div
  class="sw-empty-state"
>
  <div
    class="sw-empty-state__element"
  >
    <div
      class="sw-empty-state__icon"
    >
      <sw-icon-stub
        size="60px"
      />
    </div>
     
    <div
      class="sw-empty-state__title"
      role="alert"
    >
      
                        
                    
    </div>
     
    <div
      class="sw-empty-state__description"
    >
      <div
        class="sw-empty-state__description-content"
      />
    </div>
     
    <!---->
  </div>
</div>
`;
